### PR TITLE
Fetch ETH gas price from explorer if not provided

### PR DESCRIPTION
**Pull request by YBadiss**

### CODE REVIEW

File: src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala [hunk 1]
```diff
@@ -64,11 +58,32 @@ class ApiClient(implicit val ec: ExecutionContext) {
...
       mapper.readValue(response.contentString, classOf[FeeInfo])
     }.asScala
   }
 
+  def getGas(currencyName: String): Future[GasInfo] = {
+    val request = Request(Method.Get, "/blockchain/v3/fees")
+    val host = DaemonConfiguration.explorerApiAddresses.getOrElse(currencyName, ConfigurationDefaults.BLOCKCHAIN_DEFAULT_API_ENDPOINT)
+    request.host = host
+
+    val service = client.newService(s"$host:$port")
```
Comment (tt810) [inline src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: you are creating a service every time this method is called. did you check if that causes problems?

Comment (YBadiss) [reply src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: I was assuming that's what the session pool was for, but I can't find any proper doc on that from finagle... Unless you know it more, I'll go with the safe route of one client for eth, one for eth_ropsten, and one for btc

File: src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala [hunk 2]
```diff
@@ -25,19 +26,13 @@ import scala.concurrent.{ExecutionContext, Future}
   *
   */
 @Singleton
-class ApiClient(implicit val ec: ExecutionContext) {
+class ApiClient(implicit val ec: ExecutionContext) extends Logging {
   import ApiClient._
-  private[this] val (host, port, poolSize) = DaemonConfiguration.apiConnection
+  private[this] val (btcHost, port, poolSize) = DaemonConfiguration.apiConnection
```
Comment (tt810) [inline src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: could you refactor a bit (or if you don't have time I will do it) to combine `apiConnection` and `explorerApiAddresses`?

File: src/main/scala/co/ledger/wallet/daemon/models/Pool.scala [hunk 3]
```diff
@@ -106,9 +106,9 @@ class Pool(private val coreP: core.WalletPool, val id: Long) extends Logging {
     coreP.getCurrency(currencyName).flatMap { coreC =>
       val walletConfig = core.DynamicObject.newInstance()
       val apiUrl = DaemonConfiguration.explorerApiAddresses.getOrElse(coreC.getName, ConfigurationDefaults.BLOCKCHAIN_DEFAULT_API_ENDPOINT)
-      walletConfig.putString("BLOCKCHAIN_EXPLORER_API_ENDPOINT", apiUrl)
+      walletConfig.putString("BLOCKCHAIN_EXPLORER_API_ENDPOINT", s"http://$apiUrl")
```
Comment (tt810) [inline src/main/scala/co/ledger/wallet/daemon/models/Pool.scala]: Please don't prepend the `http://` here.

File: src/main/scala/co/ledger/wallet/daemon/models/Pool.scala [hunk 4]
```diff
@@ -106,9 +106,9 @@ class Pool(private val coreP: core.WalletPool, val id: Long) extends Logging {
...
       val wsUrl = DaemonConfiguration.explorerWebsocketAddresses.getOrElse(coreC.getName, ConfigurationDefaults.BLOCKCHAIN_OBSERVER_WS_ENDPOINT)
-      walletConfig.putString("BLOCKCHAIN_OBSERVER_WS_ENDPOINT", wsUrl)
+      walletConfig.putString("BLOCKCHAIN_OBSERVER_WS_ENDPOINT", s"ws://$wsUrl")
```
Comment (tt810) [inline src/main/scala/co/ledger/wallet/daemon/models/Pool.scala]: Please don't prepend `ws://` here

File: src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala (hunk 2 above)
Comment (YBadiss) [reply src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: this is the only place where they mean the same thing. I have done a refactor so here we only use apiConnection, but explorerApiAddresses is still used in other locations

File: src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala [hunk 5]
```diff
@@ -27,18 +27,14 @@ import scala.concurrent.{ExecutionContext, Future}
...
-    Http.client
-      .withSessionPool.maxSize(poolSize)
-      .configured(Transporter.HttpProxy(Some(new InetSocketAddress(proxyHost, proxyPort)), None))
-      .newService(s"$host:$port")
-  } else {
-    Http.client
-      .withSessionPool.maxSize(poolSize)
-      .newService(s"$host:$port")
+  private[this] val defaultHostService = {
+    val (host, port, poolSize) = DaemonConfiguration.apiConnection.default("")
```
Comment (tt810) [inline src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: I think we should be able to remove this `apiConnection.default`

File: src/main/scala/co/ledger/wallet/daemon/configurations/DaemonConfiguration.scala [hunk 6]
```diff
@@ -79,4 +76,12 @@ object DaemonConfiguration {
 
   val explorerWebsocketAddresses: scala.collection.Map[String, String] =
     config.getObject("explorer.ws").unwrapped().asScala.mapValues(_.toString)
+
+  val apiConnection: scala.collection.immutable.Map[String, (String, Int, Int)] = {
```
Comment (tt810) [inline src/main/scala/co/ledger/wallet/daemon/configurations/DaemonConfiguration.scala]: I think maybe my comments were confusing. Let's talk in person

File: src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala [hunk 7]
```diff
@@ -63,12 +57,33 @@ class ApiClient(implicit val ec: ExecutionContext) {
       case _ => throw new UnsupportedOperationException(s"currency not supported '$currencyName'")
     }
     val request = Request(Method.Get, path)
-    request.host = "api.ledgerwallet.com"
-    client(request).map { response =>
+    val (host, service) = hostServices.getOrElse(currencyName, hostServices("default"))
+    println(s"USING HOST SERVICE $currencyName => $host")
```
Comment (tt810) [inline src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: could you remove this println please?

Comment (YBadiss) [reply src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: ah forgot it 👍

File: src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala [hunk 8]
```diff
@@ -27,17 +27,11 @@ import scala.concurrent.{ExecutionContext, Future}
...
-      .configured(Transporter.HttpProxy(Some(new InetSocketAddress(proxyHost, proxyPort)), None))
-      .newService(s"$host:$port")
-  } else {
-    Http.client
-      .withSessionPool.maxSize(poolSize)
-      .newService(s"$host:$port")
+  private[this] val hostServices = DaemonConfiguration.explorerApiAddresses.map { case (currencyName, (host, port, poolSize)) =>
+    // This is required because Http.Client.newService fails if the url starts with the protocol
+    val hostNoProtocol = host.replaceFirst(".+?://", "")
+    (currencyName, (hostNoProtocol, buildClient(poolSize).newService(s"$hostNoProtocol:$port")))
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/clients/ApiClient.scala]: Here, I really don't see the need of creating a new client with a connection pool for every coin.

File: src/main/scala/co/ledger/wallet/daemon/models/Pool.scala [hunk 9]
```diff
@@ -105,9 +105,14 @@ class Pool(private val coreP: core.WalletPool, val id: Long) extends Logging {
   def addWalletIfNotExist(walletName: String, currencyName: String): Future[core.Wallet] = {
     coreP.getCurrency(currencyName).flatMap { coreC =>
       val walletConfig = core.DynamicObject.newInstance()
-      val apiUrl = DaemonConfiguration.explorerApiAddresses.getOrElse(coreC.getName, ConfigurationDefaults.BLOCKCHAIN_DEFAULT_API_ENDPOINT)
+      val apiUrl = DaemonConfiguration.explorerApiAddresses.get(coreC.getName) match {
+        case Some((host, _, _)) => host
```
Comment (luchenyuxx) [inline src/main/scala/co/ledger/wallet/daemon/models/Pool.scala]: There is port in the config, but it's not taken into account here.

**Pull request merged**